ggml/src/ggml-opt.cpp

```diff
@@ -492,23 +531,41 @@ static void ggml_opt_build(ggml_opt_context_t opt_ctx) {
     // gb_opt == graph backward optimize, forward pass, then backward pass to calculate gradients, then optimizer step.
     opt_ctx->gb_opt = ggml_graph_dup(opt_ctx->ctx_compute, opt_ctx->gb_grad, /*force_grads =*/ true);
 
-    opt_ctx->adamw_params = ggml_new_tensor_1d(opt_ctx->ctx_cpu, GGML_TYPE_F32, 7);
-    ggml_set_input(opt_ctx->adamw_params);
-    ggml_set_name(opt_ctx->adamw_params, "adamw_params");
+    opt_ctx->adamw_params = ggml_new_tensor_1d(opt_ctx->ctx_cpu, GGML_TYPE_F32, 8);
+    ggml_tensor * adamw_params = opt_ctx->adamw_params;
+    ggml_set_input(adamw_params);
+    ggml_set_name(adamw_params, "adamw_params");
 
+    std::string step_prefix     = moment ? "AdamW step for " : "SGD step for ";
+    unsigned    step_prefix_len_orig = step_prefix.size();
     for (int i = opt_ctx->gf->n_nodes-1; i >= 0; --i) {
         struct ggml_tensor * node = opt_ctx->gb_opt->nodes[i];
         struct ggml_tensor * grad = ggml_graph_get_grad(opt_ctx->gb_opt, node);
 
         if (grad && (node->flags & GGML_TENSOR_FLAG_PARAM)) {
-            struct ggml_tensor * m        = opt_ctx->grad_m[i];
-            struct ggml_tensor * v        = opt_ctx->grad_v[i];
-            struct ggml_tensor * opt_step = ggml_opt_step_adamw(opt_ctx->ctx_compute, node, grad, m, v, opt_ctx->adamw_params);
-
-            ggml_set_name(m,        (std::string("AdamW m for ")    + std::string(node->name)).c_str());
-            ggml_set_name(v,        (std::string("AdamW v for ")    + std::string(node->name)).c_str());
-            ggml_set_name(opt_step, (std::string("AdamW step for ") + std::string(node->name)).c_str());
-
+            struct ggml_tensor * m = 0;
+            struct ggml_tensor * v = 0;
+            if (moment) {
+                m = opt_ctx->grad_m[i];
+                v = opt_ctx->grad_v[i];
+                ggml_set_name(m, (std::string("AdamW m for ") + std::string(node->name)).c_str());
+                ggml_set_name(v, (std::string("AdamW v for ") + std::string(node->name)).c_str());
+            }
+            struct ggml_tensor * opt_step;
+            switch (opt_pars.optimizer) {
+                case GGML_OPT_OPTIMIZER_ADAMW:
+                    opt_step = ggml_opt_step_adamw(opt_ctx->ctx_compute, node, grad, m, v, adamw_params);
+                    break;
+                case GGML_OPT_OPTIMIZER_SGD:
+                    opt_step = ggml_opt_step_sgd(opt_ctx->ctx_compute, node, grad, adamw_params);
+                    break;
+                default:
+                    GGML_ABORT("fatal error");
+                    break;
+            }
+            step_prefix.resize(
+                step_prefix_len_orig);  // to avoid recreating a new step_prefix string temp n_nodes times
+            ggml_set_name(opt_step, (step_prefix += node->name).c_str());
```

Comment:
```suggestion
            step_prefix.resize(step_prefix_len_orig);  // to avoid recreating a new step_prefix string temp n_nodes times
            step_prefix += node->name;
            ggml_set_name(opt_step, step_prefix.c_str());
```